src/braket/experimental/autoqasm/api.py

```diff
@@ -81,26 +84,37 @@ def gate(*args) -> Callable[[Any], None]:
         Callable[[Any],]: A callable which can be used as a custom gate inside an
         aq.function or inside another aq.gate.
     """
-    return _function_wrapper(args[0], _convert_gate)
+    return _function_wrapper(args, _convert_gate)
 
 
 def _function_wrapper(
-    f: Callable,
+    args: Tuple[Any],
     converter_callback: Callable,
     converter_args: Optional[Dict[str, Any]] = None,
 ) -> Callable[[Any], aq_program.Program]:
     """Wrapping and conversion logic around the user function `f`.
 
     Args:
-        f (Callable): The target function to be converted which represents
-            the entry point of the quantum program.
+        args (Tuple[Any]): The arguments to the decorated function.
         converter_callback (Callable): The function converter, e.g., _convert_main.
         converter_args (Optional[Dict[str, Any]]): Extra arguments for the function converter.
 
     Returns:
         Callable[[Any], Program]: A callable which returns the converted
         quantum program when called.
     """
+    if not args:
+        # This the case where a decorator is called with only keyword args, for example:
+        #     @aq.main(num_qubits=4)
+        #     def my_function():
+        # To make this work, here we simply return another wrapper function which expects
+        # a Callable as the first argument.
+        def _function_wrapper_with_params(*args) -> Callable[[Any], aq_program.Program]:
+            return _function_wrapper(args, converter_callback, converter_args=converter_args)
```

Comment:
Part of why it is defined here is that it captures `converter_callback` and `converter_args`. I could move it out, but I'd have to pass these args explicitly, and then do something similar again inside that function... not sure it would really help that much? Open to ideas.

Reply:
Found a way to do this without defining a function at all, see here: https://github.com/aws/amazon-braket-sdk-python/commit/7949484c9c7af999b67f910fea05c74dfc0c7d56